Make the current loadbalancer into 'legacy'.

baldersheim opened the pull request:
Add a new 'adaptive' loadbalancer that uses select-best-of-2-random-picks.

@vekterli PR

baldersheim committed:
Make the current loadbalancer into 'legacy'.
Add a new 'adaptive' loadbalancer that uses select-best-of-2-random-picks.

baldersheim committed:
Add test and fix loadbalancer selection.

baldersheim committed:
Control docproc loadbalancer type with feature flag.

vekterli commented on documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/LoadBalancerPolicy.java:
```diff
@@ -48,12 +48,19 @@ private LoadBalancerPolicy(Map<String, String> params) {
         pattern = cluster + "/*/" + session;
-        loadBalancer = new LoadBalancer(cluster);
+        String type = params.get("type");
+        if (type == "adaptive") {
```
Should use `equals()` instead of `==`.

Also consider flipping argument order to implicitly handle `type == null` in case no balancer type is present.

vekterli replied:
Nevermind, I see this was fixed prior to me submitting this 🙂

vekterli reviewed (commented):
See comment on string comparisons, otherwise just minor comments and questions

baldersheim commented on documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/LegacyLoadBalancer.java:
```diff
@@ -0,0 +1,93 @@
+        return selectedNode;
+    }
+
+    protected NodeMetrics createNodeMetrics() {
```
Fixed

baldersheim commented on documentapi/src/test/java/com/yahoo/documentapi/messagebus/protocol/LoadBalancerTestCase.java:
```diff
@@ -39,8 +41,77 @@ private static void assertIllegalArgument(String clusterName, String recipient,
+        }
+        assertEquals(9999, sentSum);
+        assertTrue(200 > Math.abs(sumPending -  2700));
+        assertTrue( 100 > Math.abs(weights.get(0).sent() - 1780));
```
Fixed

baldersheim commented on documentapi/src/test/java/com/yahoo/documentapi/messagebus/protocol/LoadBalancerTestCase.java:
```diff
@@ -39,8 +41,77 @@ private static void assertIllegalArgument(String clusterName, String recipient,
+
+        long sentSum = 0;
+        for (var metrics : weights) {
+            assertTrue(10 > Math.abs(metrics.sent() - 3333));
```
Fixed, I also changed the test to use equals as the results are repeatable.

baldersheim commented on documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/AdaptiveLoadBalancer.java:
```diff
@@ -0,0 +1,55 @@
+            if (choices.size() > 2) {
+                candidateA = random.nextInt(choices.size());
+                candidateB = random.nextInt(choices.size());
+            }
```
You tell me, will it have any statistical impact, or will it just drown in randomness.

vekterli replied:
Probably the latter, although I'm not smart enough to crunch the numbers on it 🙂 Easiest would be to measure it in practice in the unit test over a set of runs. If there is a measurable better distribution if the two candidates are forced to be unequal, it would make sense to include it.

baldersheim replied:
After a reading a bit on the algorithm I concluded that it would indeed be better to ensure 2 distinct candidates. The reason is that when same candidate is selected fro both A and B it is in-fact a degradation into pure random. Looking at the numbers made it a simple choice.

baldersheim committed:
Follow up after PR comments.
- Make system test deterministic.
- Add @Override and be consistent on white space.

baldersheim committed:
Ensure we can not pick the same candidate twice. This make it react a lot faster especially with few nodes.
